src/resolver.rs

```diff
@@ -202,3 +281,93 @@ impl ResolverMap {
+    use crate::BlazeSymbolizer;
+    use crate::SymbolType;
+
+
+    /// Check that we can create a path to an ELF inside an APK as expected.
+    #[test]
+    fn elf_apk_path_creation() {
+        let apk = Path::new("/root/test.apk");
+        let elf = Path::new("subdir/libc.so");
+        let path = create_apk_elf_path(apk, elf).unwrap();
+        assert_eq!(path, Path::new("/root/test.apk!/subdir/libc.so"));
+    }
+
+    /// Check that we can symbolize an address using DWARF.
+    #[test]
+    fn symbolize_apk() {
+        let test_apk = Path::new(&env!("CARGO_MANIFEST_DIR"))
+            .join("data")
+            .join("test.zip");
+
+        let mmap = Mmap::builder().exec().open(test_apk).unwrap();
+        let mmap = Rc::new(mmap);
+        let archive = zip::Archive::with_mmap(mmap.clone()).unwrap();
+        let so = archive
+            .entries()
+            .find_map(|entry| {
+                let entry = entry.unwrap();
+                (entry.path == Path::new("libtest-so.so")).then_some(entry)
+            })
+            .unwrap();
+
+        // We found the ELF shared object inside the archive. Now look
+        // up the address of the `the_answer` function inside of it.
+        let elf_parser = ElfParser::from_mmap(mmap.clone(), so.data_offset).unwrap();
+        let opts = FindAddrOpts {
+            sym_type: SymbolType::Function,
+            ..Default::default()
+        };
+        let symbols = elf_parser.find_address("the_answer", &opts).unwrap();
+        // There is only one symbol with this address in the shared
+        // object.
+        assert_eq!(symbols.len(), 1);
+        let symbol = symbols.first().unwrap();
+
+        let the_answer_addr = unsafe {
+            mmap.as_ptr()
+                .add(so.data_offset)
+                // The address as reported by ELF is just an offset for
+                // our intents and purposes, because the symbol is
+                // relative to the beginning of the file.
+                .add(symbol.address)
+        };
+
+        // Now just double check that everything worked out and the function
+        // is actually where it was meant to be.
+        let the_answer_fn =
+            unsafe { transmute::<_, extern "C" fn() -> libc::c_int>(the_answer_addr) };
+        let answer = the_answer_fn();
+        assert_eq!(answer, 42);
+
+        // Now symbolize the address we just looked up. It should be
+        // correctly mapped to the `the_answer` function within our
+        // process.
+        let srcs = [SymbolSrcCfg::Process(cfg::Process { pid: None })];
+        let symbolizer = BlazeSymbolizer::new().unwrap();
+        let results = symbolizer
+            .symbolize(&srcs, &[the_answer_addr as usize])
+            .unwrap()
+            .into_iter()
+            .flatten()
+            .collect::<Vec<_>>();
+        assert_eq!(results.len(), 1);
+
+        let result = results.first().unwrap();
+        assert_eq!(result.symbol, "the_answer");
```

Comment:
nit: check that offset is zero (i.e., address points exactly to the beginning of `the_answer`)?